src/GraphQL/Execution/ExecutionHelper.cs

```diff
@@ -72,195 +72,248 @@ public static FieldType GetFieldDefinition(ISchema schema, IObjectGraphType pare
         /// </summary>
         public static Variables GetVariableValues(Document document, ISchema schema, VariableDefinitions variableDefinitions, Inputs inputs)
         {
             var variables = new Variables();
 
             if (variableDefinitions != null)
             {
-                foreach (var v in variableDefinitions)
+                foreach (var variableDef in variableDefinitions)
                 {
+                    // find the IGraphType instance for the variable type
+                    var graphType = variableDef.Type.GraphTypeFromType(schema);
+
+                    if (graphType == null)
+                    {
+                        var error = new InvalidVariableError(variableDef.Name, $"Variable has unknown type '{variableDef.Type.Name()}'");
+                        error.AddLocation(variableDef, document);
+                        throw error;
+                    }
+
+                    // create a new variable object
                     var variable = new Variable
                     {
-                        Name = v.Name
+                        Name = variableDef.Name
                     };
 
-                    if (inputs.TryGetValue(v.Name, out var variableValue))
+                    // attempt to retrieve the variable value from the inputs
+                    if (inputs.TryGetValue(variableDef.Name, out var variableValue))
                     {
-                        variable.Value = GetVariableValue(document, schema, v, variableValue);
+                        // parse the variable via ParseValue (for scalars) and ParseDictionary (for objects) as applicable
+                        variable.Value = GetVariableValue(document, graphType, variableDef, variableValue);
                     }
-                    else
+                    else if (variableDef.DefaultValue != null)
                     {
-                        var value = GetVariableValue(document, schema, v, v.DefaultValue?.Value);
-                        if (value != null)
-                            variable.Value = value;
+                        // if the variable was not specified in the inputs, and a default literal value was specified, use the specified default variable value
+
+                        // parse the variable literal via ParseLiteral (for scalars) and ParseDictionary (for objects) as applicable
+                        variable.Value = CoerceValue(schema, graphType, variableDef.DefaultValue, variables, null).Value;
+                        variable.IsDefault = true;
                     }
+                    else if (graphType is NonNullGraphType)
+                    {
+                        ThrowNullError(variable.Name);
+                    } 
+
+                    // if the variable was not specified and no default was specified, do not add the variable to the Variables object
 
+                    // add the variable to the list of parsed variables defined for the operation
```

Comment:
I don't understand how this comment correlates to the line above.

Reply:
The other comment has been fixed.